File: RouterLib/src/main/java/com/kronos/router/interceptor/RealInterceptorChain.java
package com.kronos.router.interceptor;

import android.content.Context;
import android.os.Bundle;

import com.kronos.router.exception.RouteNotFoundException;
import com.kronos.router.model.HostParams;
import com.kronos.router.model.RouterParams;

import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.Map;

public class RealInterceptorChain implements Interceptor.Chain {

    private final List<Interceptor> interceptors;
    private final String url;
    private final Map<String, HostParams> hostMap;
    private final int index;
    private Context mContext;
    private Bundle mBundle;

    RealInterceptorChain(List<Interceptor> interceptors, String url,
                         Map<String, HostParams> hosts, int index, Context context, Bundle bundle) {
        this.interceptors = interceptors;
        this.url = url;
        this.hostMap = hosts;
        this.index = index;
        this.mContext = context;
        this.mBundle = bundle;
    }


    @Override
    public Context getContext() {
        return mContext;
    }

    @Override
    public Bundle getBundle() {
        return mBundle;
    }


    @Override
    public Map<String, HostParams> getHostParams() {
        return hostMap;
    }

    @Override
    public void proceed() throws RouteNotFoundException {
        proceed(url);
    }


    public void proceed(String request) throws RouteNotFoundException {
        if (index >= interceptors.size()) {
            return;
        }

        RealInterceptorChain next = new RealInterceptorChain(interceptors, request, hostMap,
                index + 1, mContext, mBundle);
        Interceptor interceptor = interceptors.get(index);
        interceptor.intercept(next);
    }

    @NotNull
    @Override
    public String getUrl() {
        return url;
    }
}

File: RouterLib/src/main/java/com/kronos/router/interceptor/RealInterceptorChain.kt
package com.kronos.router.interceptor

import android.content.Context
import android.os.Bundle
import com.kronos.router.KRequest
import com.kronos.router.exception.RouteNotFoundException
import com.kronos.router.model.HostParams

class RealInterceptorChain internal constructor(private val interceptors: List<Interceptor>, override val url: KRequest,
                                                override val hostParams: Map<String, HostParams>?,
                                                private val index: Int, override
                                                val context: Context) : Interceptor.Chain {

    @Throws(RouteNotFoundException::class)
    override fun proceed() {
        proceed(url)
    }

    @Throws(RouteNotFoundException::class)
    fun proceed(request: KRequest) {
        if (index >= interceptors.size) {
            return
        }
        val next = RealInterceptorChain(interceptors, request, hostParams,
                index + 1, context)
        val interceptor = interceptors[index]
        interceptor.intercept(next)
    }

}